core/src/forwarding_stage.rs

```diff
@@ -38,73 +49,267 @@ mod packet_container;
 /// this should be evaluated with new stage.
 const FORWARD_BATCH_SIZE: usize = 128;
 
-/// Addresses to use for forwarding.
-#[derive(Clone)]
-pub struct ForwardingAddresses {
-    /// Forwarding address for TPU (non-votes)
-    pub tpu: Option<SocketAddr>,
-    /// Forwarding address for votes
-    pub tpu_vote: Option<SocketAddr>,
-}
-
-pub trait ForwardAddressGetter: Send + Sync + 'static {
-    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses;
+// One function has been split into two so that VoteClient and NonVoteClient
+// could manage their addresses internally, providing better encapsulation.
+pub trait ForwardAddressGetter: Clone + Send + Sync + 'static {
+    // Why added `max_count` -- because tpu-client-next refreshes connections
+    // before using them by trying to connect, while ConnectionCache refreshes
+    // them separately in a different thread by sending empty packets.
+    fn get_non_vote_forwarding_addresses(
+        &self,
+        max_count: u64,
+        protocol: Protocol,
+    ) -> Vec<Option<SocketAddr>>;
```

Comment:
why a `Vec<Option<>>`? why not just return a `Vec<>` with the size of ones you could get?

Reply:
Because the current logic design with the idea that it might be no valid leader and in this case we do:
```rust
        // Get forwarding addresses otherwise return now.
        if !self.vote_client.update_address() || !self.non_vote_client.update_address() {
            return;
        };
```

I don't know how often this happens if ever.